### ISSUE TITLE
[data grid] How to use `apiRef` to hide rows

### ISSUE THREAD

**Issue by nickfreemandesign** [mui/mui-x#13770]
### The problem in depth

TLDR: Is there any way to _hide_ a row using the `apiRef`?

We have high frequency updates to our Premium DataGrid and can no longer pass in a new collection of `rows` every time. To get away from this, we [followed the suggestions](https://mui.com/x/react-data-grid/row-updates/#high-frequency) to use `apiRef` for row updates. While this does show improvements in isolation, it poses another challenge to our use case: filtering rows using filters outside the table. 

Ideally we would like some property on the `apiRef` exposed so we can do something like:

```TS
  React.useEffect(() => {
    const subscription = interval(10).subscribe(() => {
      apiRef.current.updateRows([
        {
          id: randomInt(1, 4),
          username: randomUserName(),
          age: randomInt(10, 80),
          hidden: true
        },
        {
          id: randomInt(1, 4),
          username: randomUserName(),
          age: randomInt(10, 80),
          hidden: true 
        },
      ]);
    });
```

### Your environment

<details>
  <summary>`npx @mui/envinfo`</summary>

```
  Don't forget to mention which browser you used.
  Output from `npx @mui/envinfo` goes here.
```
</details>

**Search keywords**: apiRef hide rows
**Order ID**: 72370

**Comment by romgrk** [comment]
The grid has no concept of hidden rows beyond what it filters itself, so you can't pass a `hidden` flag. You can use the same method to delete rows from the grid:

```javascript
apiRef.current.updateRows([{ id: 1, _action: 'delete' }]);
```

But if you're changing a large number of rows, I think it's preferable to just update the `columns`. If you need further assistance make sure to provide more details for your use-case and if possible a runnable codesandbox.

**Comment by nickfreemandesign** [comment]
Hey @romgrk thanks so much for the reply. Deleting rows in place of a global filter means we need to maintain the table state somewhere else, which we were trying to avoid since its a large data set that gets streamed in. I'm not sure how updating `columns` helps me filter rows, but curious to learn more if you wouldn't mind elaborating. 

I can put together a sandbox over the weekend, but heres a diagram in case that helps in the interim:

Note the chips `all` `non-comliant` `allocations` live outside the datagrid. What we do currently is provide the data grid with the number of rows needed to start, then use the `apiRef.current.updateRows` to populate the row data as it streams through a socket connection. 

If a user clicks on any of the chips, we originally would pass in a new `rows` argument, but that just rebuilds the entire table and requires table state to live in two places.

![image](https://github.com/mui/mui-x/assets/12129271/40e6b76c-bd85-448b-8ef2-57b82b1f92a6)

**Comment by nickfreemandesign** [comment]
Also wanted to add that we have a Search Bar outside the grid as well, so a user can simply type and filter rows. All of this is of course outside the table component

**Comment by arminmeh** [comment]
@nickfreemandesign 
is the hiding of the rows only driven by clicking on chips and changing the value of the quick filter or something else can affect it as well

**Comment by nickfreemandesign** [comment]
@arminmeh Excluding actions within the table that could mutate a row (which is handled fine with `apiRef.current.updateRows`) its just clicking the chips and typing in a search bar that would also filter the rows

**Comment by arminmeh** [comment]
I assume you have a property in the model that you are targeting with the chips filters
Can you use these features
[High frequency row updates](https://mui.com/x/react-data-grid/row-updates/#high-frequency)
[Controlled filters](https://mui.com/x/react-data-grid/filtering/#controlled-filters)
[Quick filter outside of the grid](https://mui.com/x/react-data-grid/filtering-recipes/#quick-filter-outside-of-the-grid)
together to do the filtering (hiding) and keep the updates via `updateRows`

**Comment by nickfreemandesign** [comment]
I was not aware of the `quickFilterExcludeHiddinColumns` so this is helpful. If I'm pickin up what you're puttin down, then the proposed approach is to add hidden columns that are still filterable from outside the grid using the `quickFilterExcludeHiddinColumns` property. And this would play nice with not needing to pass in new rows

**Comment by arminmeh** [comment]
@nickfreemandesign 
yes, my idea was to use hidden column to filter out the rows (using existing filter features) and use update row to update the rows that are visible.
Hopefully, it can play nice performance wise

**Comment by romgrk** [comment]
Now that I think about it, you could also use [custom filter operators](https://mui.com/x/react-data-grid/filtering/customization/#create-a-custom-operator) to add your own filtering logic to the grid. It's modelized to attach a filter to a column, but you have access to all the row information so you can filter however you want.
